Playwright: refactor `EditorSettingsSidebarComponent` to use FrameLocator.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR refactors the EditorSettingsSidebarComponent to work with `FrameLocator`.

Key changes:
- change EditorSettingsSidebarComponent to accept a FrameLocator instead of Frame.
- modify CSS selectors and selector names where necessary.
- refactor methods to perform piecemeal actions in EditorSettingsSidebarComponent instead, relying on the calling class (in this case GutenbergEditorPage) to handle piecing together the actions.

#### Testing instructions

Ensure the following:
  - [x] Gutenberg E2E (desktop)
  - [x] Gutenberg E2E (mobile)
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release E2E
  - [x] i18n E2E

Related to #60868

worldomonation committed:
Begin refactor of EditorSettingsSidebarComponent.
- replace use of Frame with FrameLocator.
- update selector and selector names as necessary.
- refactor method logic.

worldomonation committed:
Fix specs to use the newly refactored EditorSettingsSidebarComponent.

worldomonation committed:
Remove pause

worldomonation committed:
Remove dead code

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -110,21 +196,21 @@ export class EditorSettingsSidebarComponent {
 	 * 	must be set if the `visibility` parameter is set to `Password`.
 	 */
-	async setVisibility(
+	async selectVisibility(
```
This method used to handle all of the process involved in setting the privacy option for an article.

In keeping with the current ethos of the refactor, the scope of this method has been reduced to handle just the interaction with the visibility popover.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -358,17 +365,100 @@ export class GutenbergEditorPage {
+		visibility: PrivacyOptions,
+		{ password }: { password?: string }
+	): Promise< void > {
+		await Promise.race( [
+			this.editorSettingsSidebarComponent.clickTab( 'Page' ),
+			this.editorSettingsSidebarComponent.clickTab( 'Post' ),
+		] );
```
You'll see this construct used throughout this PR. It's meant to make the methods that perform a series of actions to be page/post agnostic. 

The alternative was to make the calling function (which would be the test step in this case) explicitly activate the tab prior to executing actions, but I feel it doesn't make sense to do so from both a usability and logical sense.

worldomonation commented on test/e2e/specs/editor/editor__page-basic-flow.ts:
```diff
@@ -49,30 +47,30 @@ describe( DataHelper.createSuiteTitle( 'Editor: Basic Post Flow' ), function ()
-		const pageTemplateComponent = new PageTemplateModalComponent( editorIframe, page );
-		await pageTemplateComponent.selectTemplateCatagory( pageTemplateCategory );
-		await pageTemplateComponent.selectTemplate( pageTemplateLable );
+		const pageTemplateModalComponent = new PageTemplateModalComponent( editorIframe, page );
+		await pageTemplateModalComponent.selectTemplateCatagory( pageTemplateCategory );
+		await pageTemplateModalComponent.selectTemplate( pageTemplateLable );
```
This is just a rename.

worldomonation commented:
Need to fix the block inserter selector for i18n tests.

worldomonation committed:
Miscellaneous fixes:
- add handling for when non-mobile viewports try to close the sidebar.
- restore i18n compatible selector for blockInserterButton.
- remove unnecessary call to `clickSettingsTab`.

worldomonation committed:
Restore pure CSS button for settingsButton to work with i18n.

worldomonation commented:
i18n fixed and ready to review.

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -41,38 +45,50 @@ const selectors = {
+	/**
+	 * Closes the sidebar only for Mobile viewport.
+	 */
+	async closeSidebarForMobile(): Promise< void > {
+		if ( envVariables.VIEWPORT_NAME !== 'mobile' ) {
+			return;
+		}
+
+		const locator = this.frameLocator.locator( selectors.mobileCloseSidebarButton );
+		await locator.click();
```
I've moved the method here from `GutenbergEditorPage`, since the button itself is on the `EditorSettingsSidebarComponent`.

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -5,7 +5,7 @@ export type PreviewOptions = 'Desktop' | 'Mobile' | 'Tablet';
 const selectors = {
 	// Block Inserter
-	blockInserterButton: `${ panel } button[aria-label="Toggle block inserter"]`,
+	blockInserterButton: `${ panel } button.edit-post-header-toolbar__inserter-toggle`,
```
This is for i18n reasons - prior to the refactor the selector used was as shown above, but due to my changes in #61502 I ended up breaking i18n tests in `trunk`.

worldomonation commented on packages/calypso-e2e/src/lib/components/editor-toolbar-component.ts:
```diff
@@ -27,7 +27,7 @@ const selectors = {
 
 	// Editor settings
-	settingsButton: `${ panel } button[aria-label="Settings"]`,
+	settingsButton: `${ panel } .edit-post-header__settings .interface-pinned-items button:first-child`,
```
Same reasoning as above here. This fixes my error from #61502.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -579,4 +639,53 @@ export class GutenbergEditorPage {
+	 * The resulting page can change based on where you come from, and the viewport. Either way, the resulting landing spot
+	 * will have access to the Calyspo sidebar, allowing navigation around Calypso.
+	 */
+	async exitEditor(): Promise< void > {
```
This is merely a reorganization of the file, nothing changed with this method.

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -579,4 +639,53 @@ export class GutenbergEditorPage {
+	 *
+	 * @returns True if there are block warnings/errors, false otherwise.
+	 */
+	async editorHasBlockWarnings(): Promise< boolean > {
```
As above, this is an reorganization of the method order and nothing was changed.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -13,25 +19,23 @@ interface Schedule {
-	revisionsToggle: `${ sidebarParentSelector } .components-panel__body:has-text("Revisions")`,
-	lastSection: `${ sidebarParentSelector } .components-panel__body >> nth=-1`,
+	// General section-related
+	section: ( name: ArticleSections ) => `${ panel } button:has-text("${ name }")`,
```
Is only looking for buttons with text possibly too generic? Like, I'm thinking of the case of "Featured image", where both the section name and the button within there has that text in there. I think we might want to use something here to specify these to be just the section headers!

worldomonation replied:
Sure, let's try narrowing this a bit more.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -82,25 +98,100 @@ export class EditorSettingsSidebarComponent {
+		const locator = this.frameLocator.locator( selectors.tabButton( tabName ) );
+		await locator.click();
+
+		const classAttributes = await locator.getAttribute( 'class' );
+		if ( ! classAttributes?.includes( 'is-active' ) ) {
+			throw new Error( `Failed to verify ${ tabName } is active.` );
+		}
```
While I really like the simplicity of this, I'm a little bit nervous about race conditions herer, which was the original reason for using the `waitFor` combined with a selector that identifies the active state.

When you click on a tab, it's async updating the React state, which is then re-rendering that tab with the new class as calculated from that state. While that should be lightening fast, Playwright is also really fast, and `getAttribute` returns immediately. 

It's definitely possible that this race condition will never occur in reality! But I feel like we've seen similar things happen, and it makes me a little nervous.

Thoughts?

worldomonation replied:
Replaced this logic with the following:

```
		const activeTabLocator = this.frameLocator.locator(
			`${ selectors.tabButton( tabName ) }.is-active`
		);
		await activeTabLocator.waitFor();
```

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -82,25 +98,100 @@ export class EditorSettingsSidebarComponent {
-	async expandSection( sectionName: EditorSidebarSection ): Promise< void > {
-		// Avoid the wpcalypso/staging banner
+	async expandSection( name: ArticleSections ): Promise< void > {
+		if ( await this.targetIsOpen( selectors.section( name ) ) ) {
+			return;
+		}
```
I think this check works really well here, because it's not following any direct action, so I don't think we'll ever hit a race condition!

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -82,25 +98,100 @@ export class EditorSettingsSidebarComponent {
+		const sectionLocator = this.frameLocator.locator( selectors.section( name ) );
+		await sectionLocator.click();
+
+		const newState = await this.targetIsOpen( selectors.section( name ) );
```
Same question here as above... Since this is happening right after an action, are we worried about race conditions where we check the state faster than the React component can re-render?

worldomonation replied:
Replaced with `waitFor`.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -82,25 +98,100 @@ export class EditorSettingsSidebarComponent {
+		await sectionLocator.click();
+
+		const newState = await this.targetIsOpen( selectors.section( name ) );
+		if ( ! newState ) {
```
Tiny optional style nit: but with boolean variables, how do you feel about variable names that declare something that could be true or false? That is, something like `sectionIsOpen` instead of `newState`. I personally think that makes it easier to read the code flow! But, that's just my personal take, and definitely not required.

worldomonation replied:
Got it, replacing it with variable `expanded` which is a lot more intuitive.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -82,25 +98,100 @@ export class EditorSettingsSidebarComponent {
+		const buttonLocator = this.frameLocator.locator( selectors.visibilityButton );
+		await buttonLocator.click();
+
+		const newState = await buttonLocator.getAttribute( 'aria-expanded' );
```
Same question here about race conditions

worldomonation replied:
Replaced with `waitFor`.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -13,25 +19,23 @@ interface Schedule {
-	lastSection: `${ sidebarParentSelector } .components-panel__body >> nth=-1`,
+	// General section-related
+	section: ( name: ArticleSections ) => `${ panel } button:has-text("${ name }")`,
+	sectionHeader: `${ panel } .components-panel__body`,
```
I believe this is just for the revisions link, right? Should we just collapse this with the section above by keying off of the `.components-button` class? It feels a little weird having these two selectors as they seem at surface level to be referencing the same thing, and I could see folks being initially confused which they should use.

worldomonation replied:
Revisions is a bit different in that it's an `<a>` tag.

I've changed this selector to point specifically to the revisions button instead.

worldomonation replied:
So I remember why this selector was here, it was to handle the scrolling of the sidebar to the end.

worldomonation replied:
I've changed the `scrollToBottomOfSidebar` method to scroll to the `Discussion` panel, which is the last item on the page.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -134,115 +225,125 @@ export class EditorSettingsSidebarComponent {
+		const locator = this.frameLocator.locator(
+			`${ selectors.sectionHeader }:has-text("Revisions")`
+		);
+		await Promise.all( [ this.page.waitForResponse( /.*diffs\?.*/ ), locator.click() ] );
```
Do we need to wait on the web request? I assume anything that is using the revisions popup will use auto-waiting before interacting with anything on the popup, so this might just add an unnecessary wait that could flake at some point in the future.

worldomonation replied:
I felt like it was acceptable for this method to perform a guarantee, but I can remove it.

worldomonation replied:
Removed the Promise.all and response check.

dpasque commented on packages/calypso-e2e/src/lib/components/editor-settings-sidebar-component.ts:
```diff
@@ -251,28 +352,30 @@ export class EditorSettingsSidebarComponent {
+		await inputLocator.fill( slug );
+		// Hit the Tab key to confirm URL slug input and update the Post URL
+		// shown in this section.
+		await this.page.keyboard.press( 'Tab' );
+
+		const generatedURLSlugLocator = this.frameLocator.locator( selectors.permalinkGeneratedURL );
+		const generatedSlug = await generatedURLSlugLocator.innerText();
+		const expectedSlug = slug.replace( ' ', '-' );
+
+		if ( ! generatedSlug.includes( expectedSlug ) ) {
+			throw new Error(
+				`Generated URL slug ${ generatedSlug } did not match expected URL slug ${ expectedSlug }`
+			);
+		}
```
Same thing re: race conditions -- any reason to just not wait for the expected new URL? It also seems like it might be a bit simpler.

worldomonation replied:
Even in that case, we might want to narrow the selector somewhat.

worldomonation replied:
Used a bit more narrowed but simpler method of checking and it seems to work 👍

dpasque reviewed (changes requested):
Wow, this improves the readability of these files a lot!! 🎉  Thanks for doing these changes!

I have a few comments and questions. Most of the comments are around the same design question: how to verify the state of something right after we take an action. I can spin up a thread in slack to discuss that there and then we can follow-up with our decision here!

worldomonation committed:
Address PR comments:
- narrow the section selector.
- remove `sectionHeading` selector.
- use `waitFor` instead of checking attributes.
- simplify slug check.

dpasque reviewed (approved):
Updates look fantastic! 🎉  Thank you for making those!
